nextest-filtering/src/expression.rs

```diff
@@ -308,3 +295,67 @@ impl Logic for Option<bool> {
         self.map(|v| !v)
     }
 }
+
+/// Haskell madness here:
+pub enum FilteringExprLayer<'a, A> {
+    Not(A),
+    Union(A, A),
+    Intersection(A, A),
+    Set(&'a FilteringSet),
+}
+
+impl<'a, A, B> MapLayer<B> for FilteringExprLayer<'a, A> {
+    type Unwrapped = A;
+
+    type To = FilteringExprLayer<'a, B>;
+
+    #[inline(always)]
+    fn map_layer<F: FnMut(Self::Unwrapped) -> B>(self, mut f: F) -> Self::To {
+        use FilteringExprLayer::*;
+        match self {
+            Not(a) => Not(f(a)),
+            Union(a, b) => Union(f(a), f(b)),
+            Intersection(a, b) => Intersection(f(a), f(b)),
+            Set(f) => Set(f),
+        }
+    }
+}
+
+impl<'a> Project for &'a FilteringExpr {
+    type To = FilteringExprLayer<'a, &'a FilteringExpr>;
+
+    fn project(self) -> Self::To {
+        match self {
+            FilteringExpr::Not(a) => FilteringExprLayer::Not(a.as_ref()),
+            FilteringExpr::Union(a, b) => FilteringExprLayer::Union(a.as_ref(), b.as_ref()),
+            FilteringExpr::Intersection(a, b) => {
+                FilteringExprLayer::Intersection(a.as_ref(), b.as_ref())
+            }
+            FilteringExpr::Set(f) => FilteringExprLayer::Set(f),
+        }
+    }
+}
+
+// TODO: short circuit
+pub fn matches_binary_2(e: &FilteringExpr, query: &BinaryQuery<'_>) -> Option<bool> {
+    e.collapse_layers(|layer: FilteringExprLayer<Option<bool>>| {
+        match layer {
```

Comment:
Could you pick one of the two styles? Either import the variants in both cases or neither.